File: WebsiteHoTroDichVuBanDoc_Backend/WebsiteHoTroDichVuBanDoc_Backend/app/models/dat_cho_ngoi.py
from typing import Optional
from datetime import datetime
from app.models.db_base import DBModel

class DatChoNgoiBase(DBModel):
    maChoNgoi: int
    maBanDoc: int
    maNhanVien: Optional[int] = None # Nhân viên xác nhận (nếu cần)
    thoiGianBatDau: datetime
    thoiGianKetThuc: datetime
    trangThaiDatCho: str = 'kichHoat'
    thoiDiemHuy: Optional[datetime] = None
    nhanVienHuy: Optional[int] = None

class DatChoNgoiCreate(DBModel):
    # Khi user tạo, chỉ cần
    maChoNgoi: int
    maBanDoc: int
    thoiGianBatDau: datetime
    thoiGianKetThuc: datetime

class DatChoNgoiUpdate(DBModel):
    # Dùng để hủy
    trangThaiDatCho: Optional[str] = None
    thoiDiemHuy: Optional[datetime] = None
    nhanVienHuy: Optional[int] = None

class DatChoNgoi(DatChoNgoiBase):
    maDatCho: int
    ngayKhoiTao: Optional[datetime] = None

class DatChoNgoiCheckIn(DBModel):
    """
    Model input cho hành động "check-in" tại quầy.
    Chỉ cần biết nhân viên nào xử lý.
    """
    maNhanVien: int

File: WebsiteHoTroDichVuBanDoc/Backend/src/main/java/com/library/website/models/DatChoNgoi.java
package com.library.website.models;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import java.time.LocalDateTime;

@Entity
@Table(name = "datchongoi", schema = "public")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class DatChoNgoi {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "madatcho")
    private Long maDatCho;

    @ManyToOne
    @JoinColumn(name = "machongoi")
    private ChoNgoi choNgoi;

    @ManyToOne
    @JoinColumn(name = "mabandoc")
    private BanDoc banDoc;

    @Column(name = "manhanvien")
    private Long maNhanVien;

    @Column(name = "thoigianbatdau", nullable = false)
    private LocalDateTime thoiGianBatDau;

    @Column(name = "thoigianketthuc", nullable = false)
    private LocalDateTime thoiGianKetThuc;

    @Column(name = "trangthaidatcho")
    private String trangThaiDatCho = "kichHoat";

    @Column(name = "ngaykhoitao")
    private LocalDateTime ngayKhoiTao = LocalDateTime.now();

    @Column(name = "thoidiemhuy")
    private LocalDateTime thoiDiemHuy;

    @Column(name = "nhanvienhuy")
    private Long nhanVienHuy;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt = LocalDateTime.now();

    @PreUpdate
    public void preUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}
